File: common/repository-jpa/src/main/java/org/yechan/BaseEntity.java
package org.yechan;

import io.hypersistence.utils.hibernate.id.Tsid;
import jakarta.persistence.Column;
import jakarta.persistence.Id;
import jakarta.persistence.MappedSuperclass;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import java.time.LocalDateTime;
import java.util.Objects;
import org.hibernate.proxy.HibernateProxy;

@MappedSuperclass
public abstract class BaseEntity {
    @Column(updatable = false)
    private LocalDateTime createdAt;

    @Column
    private LocalDateTime updatedAt;

    @Id
    @Tsid
    private Long id;

    protected BaseEntity() {
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }

    public Long getId() {
        return id;
    }

    protected void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }

    protected void setUpdatedAt(LocalDateTime updatedAt) {
        this.updatedAt = updatedAt;
    }

    protected void setId(Long id) {
        this.id = id;
    }

    @PrePersist
    public void prePersist() {
        LocalDateTime now = LocalDateTime.now();
        createdAt = now;
        updatedAt = now;
    }

    @PreUpdate
    public void preUpdate() {
        updatedAt = LocalDateTime.now();
    }

    @Override
    public final boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        if (other == null) {
            return false;
        }

        Class<?> otherEffectiveClass = other instanceof HibernateProxy otherProxy
            ? otherProxy.getHibernateLazyInitializer().getPersistentClass()
            : other.getClass();
        Class<?> thisEffectiveClass = this instanceof HibernateProxy thisProxy
            ? thisProxy.getHibernateLazyInitializer().getPersistentClass()
            : getClass();
        if (thisEffectiveClass != otherEffectiveClass) {
            return false;
        }

        BaseEntity otherEntity = (BaseEntity) other;
        return id != null && Objects.equals(id, otherEntity.id);
    }

    @Override
    public final int hashCode() {
        return this instanceof HibernateProxy proxy
            ? proxy.getHibernateLazyInitializer().getPersistentClass().hashCode()
            : getClass().hashCode();
    }
}

File: common/repository-jpa/src/main/kotlin/org/yechan/BaseEntity.kt
package org.yechan

import io.hypersistence.utils.hibernate.id.Tsid
import jakarta.persistence.Column
import jakarta.persistence.Id
import jakarta.persistence.MappedSuperclass
import jakarta.persistence.PrePersist
import jakarta.persistence.PreUpdate
import org.hibernate.proxy.HibernateProxy
import java.time.LocalDateTime
import java.util.Objects

@MappedSuperclass
abstract class BaseEntity protected constructor() {
    @field:Column(updatable = false)
    var createdAt: LocalDateTime? = null
        protected set

    @field:Column
    var updatedAt: LocalDateTime? = null
        protected set

    @field:Id
    @field:Tsid
    var id: Long? = null
        protected set

    @PrePersist
    fun prePersist() {
        val now = LocalDateTime.now()
        createdAt = now
        updatedAt = now
    }

    @PreUpdate
    fun preUpdate() {
        updatedAt = LocalDateTime.now()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null) {
            return false
        }

        val otherEffectiveClass =
            if (other is HibernateProxy) {
                other.hibernateLazyInitializer.persistentClass
            } else {
                other.javaClass
            }
        val thisEffectiveClass =
            if (this is HibernateProxy) {
                this.hibernateLazyInitializer.persistentClass
            } else {
                javaClass
            }
        if (thisEffectiveClass != otherEffectiveClass) {
            return false
        }

        other as BaseEntity
        return id != null && Objects.equals(id, other.id)
    }

    override fun hashCode(): Int = if (this is HibernateProxy) {
        this.hibernateLazyInitializer.persistentClass.hashCode()
    } else {
        javaClass.hashCode()
    }
}
